### PR TITLE
feat(gatsby): Add JSX Runtime options to `gatsby-config.js`

### LINKED ISSUE #28657: React 17 new JSX Transform: ReferenceError: React is not defined
## Description

Gatsby 2.28.1 will throw a `ReferenceError: React is not defined` if React is not imported on top of the file, but the new JSX Transform (built-in in Gatsby since 2.24) supports files with JSX without a React Import.

### Steps to reproduce

I couldn't reproduce this in CodeSandbox but you should easily reproduce this by rendering a component that does not import React.

### Expected result

Should not throw the error and allow components without a React import to work normally

### Actual result

What happened.
**It happens in both development and build mode**
The output JS file does not have any React reference:
```javascript
__webpack_require__.r(__webpack_exports__);
/* WEBPACK VAR INJECTION */(function(__react_refresh_utils__) {/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "ExplorerTemplate", function() { return ExplorerTemplate; });
__webpack_require__.$Refresh$.runtime = __webpack_require__(/*! ./node_modules/react-refresh/runtime.js */ "./node_modules/react-refresh/runtime.js");
__webpack_require__.$Refresh$.setup(module.i);

var _jsxFileName = "/home/bartmr/Documents/unnamed-too/client-side/web-app/src/components/templates/explorer/explorer-template.tsx";
function ExplorerTemplate() {
  return /*#__PURE__*/React.createElement("div", {
    __self: this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 7,
      columnNumber: 5
    }
  });
}
_c = ExplorerTemplate;

var _c;

__webpack_require__.$Refresh$.register(_c, "ExplorerTemplate");

const currentExports = __react_refresh_utils__.getModuleExports(module.i);
__react_refresh_utils__.registerExportsForReactRefresh(currentExports, module.i);

if (true) {
  const isHotUpdate = !!module.hot.data;
  const prevExports = isHotUpdate ? module.hot.data.prevExports : null;

  if (__react_refresh_utils__.isReactRefreshBoundary(currentExports)) {
    module.hot.dispose(
      /**
       * A callback to performs a full refresh if React has unrecoverable errors,
       * and also caches the to-be-disposed module.
       * @param {*} data A hot module data object from Webpack HMR.
       * @returns {void}
       */
      function hotDisposeCallback(data) {
        // We have to mutate the data object to get data registered and cached
        data.prevExports = currentExports;
      }
    );
    module.hot.accept(
      /**
       * An error handler to allow self-recovering behaviours.
       * @param {Error} error An error occurred during evaluation of a module.
       * @returns {void}
       */
      function hotErrorHandler(error) {
        if (
          false
        ) {}

        if (typeof __react_refresh_test__ !== 'undefined' && __react_refresh_test__) {
          if (window.onHotAcceptError) {
            window.onHotAcceptError(error.message);
          }
        }

        __webpack_require__.c[module.i].hot.accept(hotErrorHandler);
      }
    );

    if (isHotUpdate) {
      if (
        __react_refresh_utils__.isReactRefreshBoundary(prevExports) &&
        __react_refresh_utils__.shouldInvalidateReactRefreshBoundary(prevExports, currentExports)
      ) {
        module.hot.invalidate();
      } else {
        __react_refresh_utils__.enqueueUpdate(
          /**
           * A function to dismiss the error overlay after performing React refresh.
           * @returns {void}
           */
          function updateCallback() {
            if (
              false
            ) {}
          }
        );
      }
    }
  } else {
    if (isHotUpdate && __react_refresh_utils__.isReactRefreshBoundary(prevExports)) {
      module.hot.invalidate();
    }
  }
}
/* WEBPACK VAR INJECTION */}.call(this, __webpack_require__(/*! ./node_modules/@pmmmwh/react-refresh-webpack-plugin/lib/runtime/RefreshUtils.js */ "./node_modules/@pmmmwh/react-refresh-webpack-plugin/lib/runtime/RefreshUtils.js")))//# sourceURL=[module]
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiLi9zcmMvY29tcG9uZW50cy90ZW1wbGF0ZXMvZXhwbG9yZXIvZXhwbG9yZXItdGVtcGxhdGUudHN4LmpzIiwic291cmNlcyI6WyIvaG9tZS9iYXJ0bXIvRG9jdW1lbnRzL3VubmFtZWQtdG9vL2NsaWVudC1zaWRlL3dlYi1hcHAvc3JjL2NvbXBvbmVudHMvdGVtcGxhdGVzL2V4cGxvcmVyL2V4cGxvcmVyLXRlbXBsYXRlLnRzeCJdLCJzb3VyY2VzQ29udGVudCI6WyJpbXBvcnQgeyBQaGFzZXJQYWdlIH0gZnJvbSAnc3JjL2NvbXBvbmVudHMvcm91dGluZy9waGFzZXItcGFnZS9waGFzZXItcGFnZSc7XG5cblxuXG5leHBvcnQgZnVuY3Rpb24gRXhwbG9yZXJUZW1wbGF0ZSgpIHtcbiAgcmV0dXJuIChcbiAgICA8ZGl2PjwvZGl2PlxuICApO1xufVxuIl0sIm1hcHBpbmdzIjoiOzs7Ozs7QUFJQTtBQUNBO0FBQ0E7QUFBQTtBQUFBO0FBQUE7QUFBQTtBQUFBO0FBQUE7QUFFQTtBQUpBO0FBQ0E7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7OztBIiwic291cmNlUm9vdCI6IiJ9
```

### Environment
**It happens in both development and build mode**
**It also happens in NodeJS 12**

  System:
    OS: Linux 4.15 Ubuntu 18.04.5 LTS (Bionic Beaver)
    CPU: (4) x64 Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
    Shell: 4.4.20 - /bin/bash
  Binaries:
    Node: 10.16.0 - ~/.nvm/versions/node/v10.16.0/bin/node
    npm: 6.9.0 - ~/.nvm/versions/node/v10.16.0/bin/npm
  Languages:
    Python: 2.7.17 - /usr/bin/python
  Browsers:
    Chrome: 87.0.4280.88
  npmPackages:
    gatsby: ^2.28.1 => 2.28.1 
    gatsby-image: ^2.7.0 => 2.7.0 
    gatsby-plugin-manifest: ^2.8.0 => 2.8.0 
    gatsby-plugin-offline: ^3.6.0 => 3.6.0 
    gatsby-plugin-react-helmet: ^3.6.0 => 3.6.0 
    gatsby-plugin-robots-txt: ^1.5.3 => 1.5.3 
    gatsby-plugin-sass: ^2.7.0 => 2.7.0 
    gatsby-plugin-sharp: ^2.10.1 => 2.10.1 
    gatsby-plugin-sitemap: ^2.8.0 => 2.8.0 
    gatsby-plugin-svgr: 2.0.2 => 2.0.2 
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.18 => 1.1.18 
    gatsby-source-filesystem: ^2.7.0 => 2.7.0 
    gatsby-transformer-sharp: ^2.8.0 => 2.8.0 
  npmGlobalPackages:
    gatsby-cli: 2.7.49

### Possible solution:
I was able to go around this bug by adding this to `gatsby-node.js`:
```javascript
exports.onCreateBabelConfig = ({ actions }) => {
  actions.setBabelPlugin({
    name: '@babel/plugin-transform-react-jsx',
    options: {
      runtime: 'automatic',
    },
  });
};
```

### LINKED ISSUE #30373: Automatic JSX runtime needs to be configured twice
I've created an example monorepo that contains Gatsby sites and components in separate packages, and uses the new automatic JSX transform in React 17.

In Gatsby v3, I seem to need to configure the JSX transform automatic runtime twice, in `gatsby-node.js` and `.babelrc.js`. This seems incorrect.

### Steps to reproduce

Minimal reproduction: https://github.com/aaronadamsCA/gatsby-example-monorepo/tree/component-library/sites/theme-ui-example

If I remove the configuration in `gatsby-node.js`, building static HTML fails with `WebpackError: ReferenceError: React is not defined`:

<details>

```
$ gatsby build
success open and validate gatsby-configs - 0.045s
warn Plugin gatsby-plugin-tsconfig-paths is not compatible with your gatsby version 3.1.1 - It requires gatsby@~2.x.x
success load plugins - 0.139s
success onPreInit - 0.037s
success initialize cache - 0.010s
success copy gatsby files - 0.260s
success onPreBootstrap - 0.039s
success createSchemaCustomization - 0.011s
success Checking for changed pages - 0.002s
success source and transform nodes - 0.138s
success building schema - 0.432s
info Total nodes: 25, SitePage nodes: 1 (use --verbose for breakdown)
success createPages - 0.003s
success Checking for changed pages - 0.001s
success createPagesStatefully - 0.109s
success update schema - 0.120s
success onPreExtractQueries - 0.002s
success extract queries from components - 0.131s
success write out redirect data - 0.002s
success onPostBootstrap - 0.007s
info bootstrap finished - 5.000s
success run static queries - 0.050s - 1/1 19.91/s
success run page queries - 0.022s - 4/4 181.94/s
success write out requires - 0.021s
success Building production JavaScript and CSS bundles - 16.162s
success Rewriting compilation hashes - 0.004s
success Building HTML renderer - 4.812s
failed Building static HTML for pages - 0.266s

 ERROR #95313 

Building static HTML failed for path "/"

See our docs page for more info on this error: https://gatsby.dev/debug-html


  4 |
  5 | export const Link: FC<{ to: string }> = (props) => (
> 6 |   <ThemedLink as={AsLink} variant="default" {...props} />
    |   ^
  7 | );
  8 |
  9 | const AsLink: FC<{ to: string }> = (props) => (


  WebpackError: ReferenceError: React is not defined
  
  - link.tsx:6 
    theme-ui-example/packages/components/src/link/link.tsx:6:3
  
  - extends.js:3 
    [theme-ui-example]/[@babel]/runtime/helpers/esm/extends.js:3:1
  
  - extends.js:9 
    [theme-ui-example]/[@babel]/runtime/helpers/esm/extends.js:9:1
  
  - extends.js:8 
    [theme-ui-example]/[@babel]/runtime/helpers/esm/extends.js:8:1
  
  - static-entry.js:263 
    theme-ui-example/.cache/static-entry.js:263:20
  


error Command failed with exit code 1.
```

</details>

If I remove the configuration in `babelrc.json`, ESLint fails to parse JavaScript files with `Parsing error: Support for the experimental syntax 'jsx' isn't currently enabled`. TypeScript files seem to be unaffected.

I would expect Gatsby and ESLint to be using the same Babel configuration.

### Notes

I created this example repository because I've found it very challenging to migrate my existing (private) Gatsby v2 monorepo to Gatsby v3. It uses mixed JS and TS, and everything works using `actions.setBabelPlugin` in `gatsby-node.js`.

In v3, since [`babel-preset-gatsby` now accepts `reactImportSource` which is passed to the underlying `@babel/preset-react` `importSource` field](https://www.gatsbyjs.com/docs/reference/release-notes/v3.0/#babel-preset-gatsby), I expected this to replace `actions.setBabelPlugin` in v3. It seems this may only be true for TypeScript components, and that JavaScript components still require the old-style Webpack configuration.

### Environment

```
  System:
    OS: Linux 4.19 Debian GNU/Linux 10 (buster) 10 (buster)
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
    Shell: 5.0.3 - /bin/bash
  Binaries:
    Node: 14.16.0 - /tmp/yarn--1616344297767-0.7405160842411349/node
    Yarn: 1.22.5 - /tmp/yarn--1616344297767-0.7405160842411349/yarn
    npm: 6.14.11 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
```

### PR DESCRIPTION
**Pull request by herecydev**
## Description

Follow up to #32972.

Allow automatic react runtime to be enabled through config, the current workaround is to start manually setting babelPlugins and eslint config. There's going to be very few people doing this for something that is supposed to work out the box

## Related Issues

Fixes #28657
Fixes #30373

### CODE REVIEW

File: docs/docs/reference/config-files/gatsby-config.md [hunk 1]
```diff
@@ -355,3 +355,13 @@ See more about [Proxying API Requests in Develop](/docs/api-proxy/).
 ## Advanced proxying with `developMiddleware`
 
 See more about [adding develop middleware](/docs/api-proxy/#advanced-proxying).
+
+## jsxAutomaticRuntime
+
+Allows using JSX without having to import React.
```
Comment (herecydev) [inline docs/docs/reference/config-files/gatsby-config.md:361]: Might need further elaboration?

File: packages/gatsby/src/joi-schemas/joi.ts [hunk 2]
```diff
@@ -49,6 +49,7 @@ export const gatsbyConfigSchema: Joi.ObjectSchema<IGatsbyConfig> = Joi.object()
       )
       .single(),
     developMiddleware: Joi.func(),
+    jsxAutomaticRuntime: Joi.boolean().default(false),
```
Comment (wardpeet) [inline packages/gatsby/src/joi-schemas/joi.ts:52]: Could we move to these 2 options please.
```suggestion
    jsxRuntime: Joi.string().default('classic'),
    jsxImportSource: Joi.string()
```

**Review by wardpeet — COMMENTED** [review]
Awesome! Thank you very much.

Could you also add:
- `"react/jsx-uses-react": "off"` to eslint-config.ts

**Comment by herecydev** [comment]
>     * "react/jsx-uses-react": "off"

Added that, it looks like the eslint rule won't scan node_modules so plugins will be able to "mix and match" without throwing eslint errors. In addition enforced that the jsxRuntime == "automatic" | "classic"

**Comment by herecydev** [comment]
Might have missed your last part of the comment

> jsxImportSource: Joi.string()

What do you want this to do?

**Comment by wardpeet** [comment]
We should also set an option to change the importSource to for example to emotion. Ideally, we want to also add an action so plugins can set it themselves.

**Comment by herecydev** [comment]
@wardpeet I think this is good to go now, `jsxRuntime` works as expected, `jsxImportSource` is present but nothing listens to it at the moment (I think another PR could take care of that, preferably someone with emotion pragma experience).

There's a failing test but I'm convinced that's unrelated.

**Comment by herecydev** [comment]
@wardpeet I'm struggling with these tests, is it usual to have this level of flake (each run through spits out some new error). I also have no means to rerun this stuff and try and get it through!

**Comment by vladar** [comment]
@wardpeet This is a new feature, not a hotfix - do we actually want to backport it to v3? If so, this likely needs a minor release, not a patch.

**Comment by merrywhether** [comment]
I've been successfully using the new transform via the `gatsby-node` method:
```js
exports.onCreateBabelConfig = ({ actions }) => {
  actions.setBabelPlugin({
    name: '@babel/plugin-transform-react-jsx',
    options: {
      runtime: 'automatic',
    },
  });
};
```

I just updated to `v4.2.0` and thought this was going to be included. I just wanted to verify that the intent is that merely adding `jsxRuntime: 'automatic'` to my `gatsby-config` should allow me to remove ^ from `gatsby-node`. Currently with ^ removed I wind up with this [older `originalFactory.call` issue](https://github.com/gatsbyjs/gatsby/discussions/30748) with or without `jsxRuntime` set, and wanted to make sure I was using it right before delving deeper into my web of plugins.

**Pull request merged**